sirepo/sim_api/jupyterhublogin.py

```diff
@@ -0,0 +1,170 @@
+def api_redirectJupyterHub():
+    is_new_user = _create_user()
+    if not cfg.rs_jupyter_migrate or not is_new_user:
+        return sirepo.http_reply.gen_redirect('jupyterHub')
+    return sirepo.http_reply.gen_json_ok()
+
+
+def logged_in_user_name(check_path=True):
+    return _user_name(sirepo.auth.logged_in_user(check_path=check_path))
+
+
+def init_apis(*args, **kwargs):
+    global cfg
+
+    if cfg:
+        return
+    cfg = pkconfig.init(
+        dst_db_root=(
+            pkio.py_path(sirepo.srdb.root()).join('jupyterhub'),
+            pkio.py_path,
+            'existing jupyter user db (ex /srv/jupyterhub)',
+        ),
+        rs_jupyter_migrate=(False, bool, 'give user option to migrate data from jupyter.radiasoft.org'),
+        src_db_root=(
+            pkio.py_path('/dev/null'),
+            pkio.py_path,
+            'new jupyter user db',
+        ),
+        uri_root=('jupyter', str, 'the root uri of jupyterhub'),
+    )
+    sirepo.auth_db.init_model(_init_model)
+    sirepo.events.register({
+        sirepo.events.Type.AUTH_LOGOUT: _event_auth_logout,
+        sirepo.events.Type.END_API_CALL: _event_end_api_call,
+        sirepo.events.Type.GITHUB_AUTHORIZED: _event_github_authorized,
+    })
+
+
+def _create_user():
+    def __user_name(logged_in_user_name):
+        n = re.sub(
+            '\W+',
+            '_',
+            logged_in_user_name.split('@')[0],
+        )
+        u = JupyterhubUser.search_by(
+            user_name=n,
+        )
+        if u:
+            n += random.choice(string.ascii_lowercase)
+        return n
+
+    with sirepo.auth_db.thread_lock:
+        if logged_in_user_name():
+            return False
+        i = sirepo.auth.logged_in_user()
+        u = sirepo.auth.logged_in_user_name()
+        assert i and u, \
+            f'Must have logged in user with user_name. uid={i} user_name={u}'
+        JupyterhubUser(
+            uid=i,
+            user_name=__user_name(u),
+        ).save()
+        return True
+
+
+def _event_auth_logout(kwargs):
+    flask.g.jupyterhub_logout_user_name = _user_name(kwargs.uid)
+
+
+def _event_end_api_call(kwargs):
+    u = flask.g.get('jupyterhub_logout_user_name', None)
+    if not u:
+       return
+    for c in (
+            ('jupyterhub-hub-login', 'hub'),
+            (f'jupyterhub-user-{u}', f'user/{u}'),
+    ):
+        kwargs.resp.delete_cookie(
+            c[0],
+            # Trailing slash is required in paths
+            path=f'/{cfg.uri_root}/{c[1]}/',
+        )
+
+
+def _event_github_authorized(kwargs):
+    if not cfg.rs_jupyter_migrate:
+        return
+    with sirepo.auth_db.thread_lock:
+        s = cfg.src_db_root.join(kwargs.user_name)
+        u = logged_in_user_name()
+        assert u, 'need logged in JupyterhubUser'
+        d = cfg.dst_db_root.join(u)
+        try:
+            s.rename(d)
+        except py.error.ENOTDIR:
+            # TODO(e-carlin): Maybe raise an error letting the user know
```

Comment:
@robnagler what do you think we should do here? Redirect them to a page saying that the move didn't find anything and ask if they want to try again?

Reply:
I think there are many failure points. I think we should just wait for people to contact us. For many users, it might be they think they had a jupyter account, and they didn't.

Reply:
3682ba2e Ok, left the logging and we can see.